File: src/platform/web/canvas.cpp
//
// Copyright © 2018 Sandcastle Software Ltd. All rights reserved.
//
// This file is part of 'Oaknut' which is released under the MIT License.
// See the LICENSE file in the root of this installation for details.
//
#if PLATFORM_WEB

#include <oaknut.h>
#include "canvas.h"


/**
 Path2D is not supported on mobile browsers, yet, hence we have to manually track Paths
 */
enum PathElementType {
    TypeMoveTo,
    TypeLineTo,
    TypeCurveTo,
};

class PathElement {
public:
    enum PathElementType type;
    PathElement(PathElementType type) {
        this->type = type;
    }
};
class PathElementMoveTo : public PathElement {
public:
    PathElementMoveTo(const POINT& pt) : PathElement(TypeMoveTo) {
        this->pt = pt;
    }
    POINT pt;
};
class PathElementLineTo : public PathElement {
public:
    PathElementLineTo(const POINT& pt) : PathElement(TypeLineTo) {
        this->pt = pt;
    }
    POINT pt;
};
class PathElementCurveTo : public PathElement {
public:
    PathElementCurveTo(const POINT& ctrl1, const POINT& ctrl2, const POINT& pt) : PathElement(TypeCurveTo) {
        this->ctrl1 = ctrl1;
        this->ctrl2 = ctrl2;
        this->pt = pt;
    }
    POINT ctrl1,ctrl2,pt;
};

class WebPath : public Path {
public:
    WebPath() {
    }
    ~WebPath() {
        for (auto it : _pathElements) {
            delete (PathElement*)it;
        }
    }
    void moveTo(POINT pt) {
        _pathElements.push_back(new PathElementMoveTo(pt));
    }
    void lineTo(POINT pt) {
        _pathElements.push_back(new PathElementLineTo(pt));
    }
    void curveTo(POINT ctrl1, POINT ctrl2, POINT pt) {
        _pathElements.push_back(new PathElementCurveTo(ctrl1, ctrl2, pt));
    }
    
    vector<PathElement*> _pathElements;
};


WebCanvas::WebCanvas() : _canvas(val::null()), _ctxt(val::null()) {
    _canvas = val::global("document").call<val>("createElement", val("canvas"));
    _bitmap = new Bitmap(this);
}

Bitmap* WebCanvas::getBitmap() {
    return _bitmap;
}

void WebCanvas::resize(int width, int height) {
    _bitmap->_width = width;
    _bitmap->_height = height;
    _canvas.set("width", width);
    _canvas.set("height", height);
    _ctxt = _canvas.call<val>("getContext", val("2d"));
    _hasChanged = true;
}
void WebCanvas::clear(COLOR color) {
    _hasChanged = true;

}
void WebCanvas::setStrokeWidth(float strokeWidth) {
    _strokeWidth = strokeWidth;
}
void WebCanvas::setStrokeColor(COLOR color) {
    _strokeColor = color;
}
void WebCanvas::setFillColor(COLOR color) {
    _fillColor = color;
}
string cssColorStr(COLOR color) {
    char ach[16];
    // CSS ordering of 32-bit hex string is RGBA
    color = ((color&0xFF000000)>>24) // i.e. move A to least significant byte
           | ((color&0xFF0000)<<8)
           | ((color&0xFF00)<<8)
           | ((color&0xFF)<<8);
    sprintf(ach, "#%08X", color);
    return string(ach);
}
void WebCanvas::setAffineTransform(AffineTransform* t) {
    _transform = t ? (*t) : AffineTransform();
    _hasChanged = true;
}
void WebCanvas::drawRect(RECT rect) {
    _ctxt.call<void>("fillRect", val(rect.left()), val(rect.top()), val(rect.size.width), val(rect.size.height));
    if (_strokeWidth > 0) {
        _ctxt.set("lineWidth", val(_strokeWidth));
        _ctxt.set("strokeStyle", val(cssColorStr(_strokeColor)));
        _ctxt.call<void>("strokeRect", val(rect.left()), val(rect.top()), val(rect.size.width), val(rect.size.height));
    }
    _hasChanged = true;
}
void WebCanvas::drawOval(RECT rect) {
    _ctxt.call<void>("save");
    _ctxt.call<void>("beginPath");
    _ctxt.call<void>("ellipse", val(rect.midX()), val(rect.midY()), val(rect.size.width/2), val(rect.size.height/2), val(0), val(0),val(2*M_PI));
    _ctxt.set("lineWidth", val(_strokeWidth));
    _ctxt.set("strokeStyle", val(cssColorStr(_strokeColor)));
    _ctxt.call<void>("stroke");
    _ctxt.call<void>("restore");
    _hasChanged = true;
}
void WebCanvas::drawPath(Path* apath) {
    WebPath* path = (WebPath*)apath;
    _ctxt.call<void>("save");
    _ctxt.call<void>("beginPath");
    _ctxt.set("lineCap", val("round"));
    _ctxt.set("lineJoin", val("round"));
    _ctxt.set("lineWidth", val(_strokeWidth));
    _ctxt.set("strokeStyle", val(cssColorStr(_strokeColor)));
    for (auto it : path->_pathElements) {
        switch (it->type) {
            case TypeMoveTo: {
                PathElementMoveTo* moveTo = (PathElementMoveTo*)it;
                POINT pt =  _transform.applyToPoint(moveTo->pt);
                _ctxt.call<void>("moveTo", val(pt.x), val(pt.y));
            }
                break;
            case TypeLineTo: {
                PathElementLineTo* lineTo = (PathElementLineTo*)it;
                POINT pt =  _transform.applyToPoint(lineTo->pt);
                _ctxt.call<void>("lineTo", val(pt.x), val(pt.y));
            }                    break;
            case TypeCurveTo: {
                PathElementCurveTo* curveTo = (PathElementCurveTo*)it;
                POINT ctrl1 =  _transform.applyToPoint(curveTo->ctrl1);
                POINT ctrl2 =  _transform.applyToPoint(curveTo->ctrl2);
                POINT pt =  _transform.applyToPoint(curveTo->pt);
                _ctxt.call<void>("bezierCurveTo", val(ctrl1.x), val(ctrl1.y), val(ctrl2.x), val(ctrl2.y), val(pt.x), val(pt.y));

            }
                break;
        }
    }
    _ctxt.call<void>("stroke");
    _ctxt.call<void>("restore");
    _hasChanged = true;
}

void WebCanvas::drawBitmap(Bitmap* bitmap, const RECT& rectSrc, const RECT& rectDst) {
    
    // Get the source JS image. If there is no JS source (i.e. bitmap only exists
    // in the C++ heap) then a temporary JS Canvas must be created for it
    val sourceImage = bitmap->_img;
    if (sourceImage.isNull()) {
        if (bitmap->_canvas) {
            sourceImage = bitmap->_canvas->_canvas;
        } else {
            val canvas = val::global("document").call<val>("createElement", val("canvas"));
            canvas.set("width", val(bitmap->_width));
            canvas.set("height", val(bitmap->_height));
            val ctxt = canvas.call<val>("getContext", val("2d"));
            val buff = val(typed_memory_view((size_t)bitmap->_pixelData.cb, (unsigned char*)bitmap->_pixelData.data));
            val imgdata = ctxt.call<val>("createImageData", val(bitmap->_width), val(bitmap->_height));
            val imgdatabuff = imgdata["data"];
            imgdatabuff.call<void>("set", buff);
            ctxt.call<void>("putImageData", imgdata, val(0), val(0));
            sourceImage = canvas;
        }
    }
    
    _ctxt.call<void>("drawImage", sourceImage,
         val(rectSrc.origin.x), val(rectSrc.origin.y), val(rectSrc.size.width), val(rectSrc.size.height),
         val(rectDst.origin.x), val(rectDst.origin.y), val(rectDst.size.width), val(rectDst.size.height));
}

Path* WebCanvas::createPath() {
    return new WebPath();
}

Canvas* Canvas::create() {
    return new WebCanvas();
}

#endif

File: src/platform/android/java/oaknut/Canvas.java
package oaknut;


import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Xfermode;

import java.nio.Buffer;

public class Canvas {

    android.graphics.Canvas canvas;
    Paint paint;
    Bitmap bitmap;
    Rect rect;
    RectF rectF;
    int fillColor;
    int strokeColor;
    float strokeWidth;
    Rect rectSrc;
    Rect rectDst;

    public Canvas() {
        canvas = new android.graphics.Canvas();
        paint = new Paint();
        rect = new Rect();
        rectSrc = new Rect();
        rectDst = new Rect();

        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setAntiAlias(true);
        //paint.setStrokeMiter();
    }

    public void resize(int width, int height) {
        if (bitmap != null) {
            canvas.setBitmap(null);
            bitmap.recycle();
        }
        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        canvas.setBitmap(bitmap);
        rect.set(0,0,width,height);
    }

    public void clear(int clearColor) {
        paint.setColor(clearColor);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC));
        canvas.drawRect(rect, paint);
        paint.setXfermode(null);
    }

    public void setFillColor(int fillColor) {
        this.fillColor = fillColor;
    }
    public void setStrokeColor(int strokeColor) {
        this.strokeColor = strokeColor;
    }
    public void setStrokeWidth(float strokeWidth) {
        this.strokeWidth = strokeWidth;
        paint.setStrokeWidth(strokeWidth);
    }


    public void setTransform(float a, float b, float c, float d, float tx, float ty) {
        Matrix matrix = new Matrix();
        float vals[] = new float[] {
               a,c,tx,b,d,ty,0,0,1
        };
        matrix.setValues(vals);
        //canvas.setMatrix(matrix);
        this.matrix = matrix;
    }
    Matrix matrix;

    public void clearTransform() {
        canvas.setMatrix(null);
    }
    public void drawRect(float x, float y, float width, float height) {
        if (fillColor != 0) {
            paint.setColor(fillColor);
            paint.setStyle(Paint.Style.FILL);
            canvas.drawRect(x,y,x+width,y+height, paint);
        }
        if (strokeColor != 0) {
            paint.setColor(strokeColor);
            paint.setStyle(Paint.Style.STROKE);
            canvas.drawRect(x,y,x+width,y+height, paint);
        }
    }
    public void drawOval(float x, float y, float width, float height) {
        if (rectF == null) {
            rectF = new RectF();
        }
        rectF.set(x,y,x+width,y+height);
        if (fillColor != 0) {
            paint.setColor(fillColor);
            paint.setStyle(Paint.Style.FILL);
            canvas.drawOval(rectF, paint);
        }
        if (strokeColor != 0) {
            paint.setColor(strokeColor);
            paint.setStyle(Paint.Style.STROKE);
            canvas.drawOval(rectF, paint);
        }
    }
    public void drawPath(Path path) {
        if (matrix != null) {
            Path path2= new Path();
            path.transform(matrix, path2);
            path = path2;
        }
        if (fillColor!=0) {
            paint.setColor(fillColor);
            paint.setStyle(Paint.Style.FILL);
            canvas.drawPath(path, paint);
        }
        if (strokeColor!=0) {
            paint.setColor(strokeColor);
            paint.setStyle(Paint.Style.STROKE);
            canvas.drawPath(path, paint);
        }
    }
    public void drawBitmap(Bitmap bitmap, float sx, float sy, float sw, float sh, float dx, float dy, float dw, float dh) {
        rectSrc.left = (int)sx;
        rectSrc.top = (int)sy;
        rectSrc.right = (int)(sx+sw);
        rectSrc.bottom = (int)(sy+sh);
        rectDst.left = (int)dx;
        rectDst.top = (int)dy;
        rectDst.right = (int)(dx+dw);
        rectDst.bottom = (int)(dy+dh);
        canvas.drawBitmap(bitmap, rectSrc, rectDst, paint);
    }

    public Bitmap getBitmap() {
        return bitmap;
    }
}
